### ISSUE TITLE
client incorrect ack matching in confirmable messages

### ISSUE THREAD

**Issue by fedepell** [obgm/libcoap#270]
While working on kYc0o/RIOT#15 it seems that the coap-client will not match correctly the IDs when receiving ACKs.

The scenario is:
- Client starts sending blocks correctly
- Server side (due to intensive local computation) will stop sending ACKs for some time
- Client will correctly resend these packets
- Server side will then ACK the packets (the original one and the resends)
- Client will match the ACKs done for the resends with next ids packets

The session is therefore desynced and this becomes a serious problem when (as it was the case) a packet is lost, as since wrong ACKs are matched there is no knowledge of a the packet being lost at all (since we have this extra ACKs in between they will match numerically lost packets).

A screenshot:

![ota-screenshot](https://user-images.githubusercontent.com/19709142/48209733-2028f000-e376-11e8-902d-300458c97d87.png)

And attached a pcapng file (a separate session from screeshot but same behaviour) of the whole session:

[ota-update-coap.zip](https://github.com/obgm/libcoap/files/2562425/ota-update-coap.zip)

There was already some ongoing discussion on the referenced kYc0o/RIOT#15 if it may be of help.

**Comment by mrdeep1** [comment]
Thanks for raising this issue.  PR #271 has been raised to fix it.  For the coap-client, it is a simple code change to use the last ack'd block number for determining the next block to send.

**Comment by kb2ma** [comment]
Thanks for the quick response! I also have been participating in the discussion on the RIOT side. Is the issue really mismatching block numbers? From the zipped pcap, it looks to me that the issue on the libcoap side is responding to a duplicate ACK for a particular message ID. For example, packet no. 21 acks message ID 29549 for a second time. I would expect that libcoap is waiting for an ACK on packet no. 20, for message ID 29550, and it should take no action based no. 21. 

I describe my perception of the issues in more detail in the linked [comment](https://github.com/kYc0o/RIOT/pull/15#issuecomment-436254481) from the RIOT thread.

**Comment by mrdeep1** [comment]
@kb2ma libcoap is responding to the duplicate ack by sending the next block (and incrementing the next block to send count), not the current block just ack'd plus one.  This is separate from the ack for a particular message id - which is symptomatic of what you are seeing - not the root cause I believe.

I suggest that you just try my suggested fix and let us know how it goes.  I have tested the fix in a lossy environment and behaves as expected in terms of recovery - but it is not the same as your file upload having a big 'think'.

As I read RFC 7959, if the server gets a duplicate data block , it should ack receipt it but not re-process the data.
It is unclear what the client should do when receiving a duplicate block ack (may have missed this in the skim read of RFC).  Should it resend the block or not?
Thinking about this, if the server knows that it has not received a specific block, it can only resend a block ack for the last block received if it wants to get more data.  The client should then re-send the ack'd block number + 1 on receipt of the duplicate block ack, so I have left the client code doing this.

**Comment by fedepell** [comment]
@mrdeep1: Thanks a lot for the very quick reaction!

I tried right now your fix. The final result is that it indeed works much better now! The pattern is a bit strange tho, and I guess this directly connects to @kbma2 comment, as now some packets will be sent many more times as a consequence of receiving the repeated ACKs from server side.
The server ignores them as already received so still final result is all correct! 

Note: At the end they get syncronized since some packets are effectively lost on the flaky connection and the numbers match again.

I attach two samples of pcap logs (well they are more or less the same) with the same transfers as last time:

[ota-after-fix-1.zip](https://github.com/obgm/libcoap/files/2565773/ota-after-fix-1.zip)

[ota-after-fix-2.zip](https://github.com/obgm/libcoap/files/2565774/ota-after-fix-2.zip)

**Comment by mrdeep1** [comment]
It is good that things ar now 'working'.

I am not expecting to see a 4.02 response (Bad Option) to the final block PUT  where M = 0.  I would be expecting a 2.01 or 2.04.  This would be a bug in the server.

In terms of re-sending the block when receiving a duplicate block ack, I am trying to find where it may be specified as the action to take (resend the next block, or just ignore the duplicate block ack).  Any pointers most welcome!

**Comment by kb2ma** [comment]
I based my comment on 7252, [5.3.2](https://tools.ietf.org/html/rfc7252#section-5.3.2), partially quoted below. The absence of a token means a zero-length token, which matches any other zero-length token. I have not looked closely at 7959, but will do that now.

The exact rules for matching a response to a request are as follows:

   1.  The source endpoint of the response MUST be the same as the
       destination endpoint of the original request.

   2.  In a piggybacked response, the Message ID of the Confirmable
       request and the Acknowledgement MUST match, and the tokens of the
       response and original request MUST match.  In a separate
       response, just the tokens of the response and original request
       MUST match.

   In case a message carrying a response is unexpected (the client is
   not waiting for a response from the identified endpoint, at the
   endpoint addressed, and/or with the given token), the response is
   rejected (Sections 4.2 and 4.3).

4.2 goes on to say:

    Rejecting an
   Acknowledgement or Reset message (including the case where the
   Acknowledgement carries a request or a code with a reserved class, or
   the Reset message is not Empty) is effected by silently ignoring it.

**Comment by mrdeep1** [comment]
If the PUT/POST request is contained within a single packet, then the response must contain a matching token.  As the PUT/POST request is now spread across multiple packets, I would expect the same token to be used for all the blocks and all the block acks (again RFC7979 is unclear about this).  However, in RFC7979, the GET example in 3.4. Combining Observe and Block2 Figure 12: shows the same token in use for the initial GET with the multiple responses (0xfb) with the subsequently triggered Observe response using a different token 0xfc.  So, I think Tokens are not at issue here.

**Comment by mrdeep1** [comment]
The closest I am getting to not re-sending the same block on receipt of a duplicate block ack.  A message is either the PUT or the ACK.

````
RFC7252 4.5 Message Deduplication.

   A recipient might receive the same Confirmable message (as indicated
   by the Message ID and source endpoint) multiple times within the
   EXCHANGE_LIFETIME (Section 4.8.2), for example, when its
   Acknowledgement went missing or didn't reach the original sender
   before the first timeout.  The recipient SHOULD acknowledge each
   duplicate copy of a Confirmable message using the same
   Acknowledgement or Reset message but SHOULD process any request or
   response in the message only once.
````
In the case of the client,
````
"The recipient SHOULD acknowledge each duplicate copy of a Confirmable 
message using the same Acknowledgement or Reset message"
````
is not appropriate, but I believe that
````
"but SHOULD process any request or response in the message only once."
````
is appropriate here to not re-sending the same block twice.  Thoughts?

**Comment by kb2ma** [comment]
On tokens, 7959, pg. 27 in sec. 3.4 also says as quoted below. I agree with you that I don't think tokens are the issue here.

```
"(Note that the choice of token 0xfc in this example is arbitrary;
   tokens are just shown in this example to illustrate that the requests
   for additional blocks cannot make use of the token of the Observation
   relationship.  As a general comment on tokens, there is no other
   mention of tokens in this document, as block-wise transfers handle
   tokens like any other CoAP exchange.  As usual, the client is free to
   choose tokens for each exchange as it likes.)"
```

**Comment by kb2ma** [comment]
I finished looking at 7959, and I found just a couple of relevant passages. They don't change the reasoning from my [original message](https://github.com/obgm/libcoap/issues/270#issuecomment-437325578) in this thread and [followup](https://github.com/obgm/libcoap/issues/270#issuecomment-437357337) quoting 7252: If the sender receives a duplicate ACK in a block exchange, it should ignore the duplicate and not send anything to the recipient based on it. 

Packet numbers below refer to the original pcap from @fedepell.

Sec 2.3, when talking about Block1 in control usage (in the ACK from the recipient, RIOT in our discussion)

```
"The NUM field of the Block1 Option indicates what block number
         is being acknowledged."
```
This reinforces that pkt no. 21 from RIOT is ACKing Block no. 3. So, it does not make sense then that libcoap sends Block no. 5 in the next packet, pkt no. 22.

Also in 2.3,
```
"Using one or both Block options, a single REST operation can be split
   into multiple CoAP message exchanges.  As specified in [RFC7252],
   each of these message exchanges uses their own CoAP Message ID."
```
This also reinforces that the Message ID is signficant. If we can't find specific direction in 7959, then the rules in 7252 apply.

**Comment by kb2ma** [comment]
Looking forward to testing the update to ignore duplicate block ACKs. Thanks for the discussion!

**Comment by mrdeep1** [comment]
#271 updated example client to ignore duplicate BLOCK1 ACKs as determined by message ID.

**Comment by fedepell** [comment]
@mrdeep1 thanks for the changes (and of course also @kb2ma for the discussion!)

Please find a log with the last modification, now it looks more like what one would expect (RFC compliant or not is another thing) without all that extra packets:

[ota-after-fix-3.zip](https://github.com/obgm/libcoap/files/2566583/ota-after-fix-3.zip)

**Comment by mrdeep1** [comment]
Yes that looks much better.  However, your server is still returning 4.02 at the end, instead of possibly 2.04 (note the dyslexia re-ordering) which needs to be fixed.

**Comment by mrdeep1** [comment]
@fedepell Now that #271 has been merged into the develop branch, can this be closed now?

**Comment by fedepell** [comment]
@mrdeep1: sure thing, closing now and many thanks again for your work!
